flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamSort.scala

```diff
@@ -0,0 +1,167 @@
+import org.apache.flink.api.common.functions.RichMapFunction
+import org.apache.flink.api.common.functions.RichFlatMapFunction
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.util.Collector
+import org.apache.flink.api.common.state.ValueState
+import org.apache.flink.api.common.state.ValueStateDescriptor
+import org.apache.flink.api.java.typeutils.RowTypeInfo
+import org.apache.flink.table.functions.ProcTimeType
+import org.apache.flink.table.functions.RowTimeType
+import org.apache.calcite.rel.core.Sort
+import org.apache.flink.api.java.functions.NullByteKeySelector
+import org.apache.calcite.rel.RelFieldCollation.Direction
+
+
+/**
+  * Flink RelNode which matches along with Sort Rule.
+  *
+  */
+class DataStreamSort(
+  calc: LogicalSort,
+  cluster: RelOptCluster,
+  traitSet: RelTraitSet,
+  inputNode: RelNode,
+  rowRelDataType: RelDataType,
+  inputType: RelDataType,
+  description: String)
+    extends SingleRel(cluster, traitSet, inputNode) with DataStreamRel {
+
+  override def deriveRowType(): RelDataType = rowRelDataType
+
+  override def copy(traitSet: RelTraitSet, inputs: java.util.List[RelNode]): RelNode = {
+    new DataStreamSort(
+      calc,
+      cluster,
+      traitSet,
+      inputs.get(0),
+      rowRelDataType,
+      inputType,
+      description + calc.getId())
+  }
+
+  override def toString: String = {
+    s"Sort($calc)"+
+      s"on fields: (${calc.collation.getFieldCollations})"
+  }
+
+  override def explainTerms(pw: RelWriter): RelWriter = {
+    super.explainTerms(pw)
+      .item("aggregate", calc)
+      .item("sort fields",calc.collation.getFieldCollations)
+      .itemIf("offset", calc.offset, calc.offset!=null)
+      .itemIf("fetch", calc.fetch, calc.fetch!=null)
+      .item("input", inputNode)
+  }
+
+  override def translateToPlan(tableEnv: StreamTableEnvironment): DataStream[Row] = {
+
+    val inputDS = getInput.asInstanceOf[DataStreamRel].translateToPlan(tableEnv)
+    
+    //need to identify time between others order fields. Time needs to be first sort element
+    val timeType = SortUtil.getTimeType(calc,inputType)
+    
+    //time ordering needs to be ascending
+    if(SortUtil.getTimeDirection(calc)!=Direction.ASCENDING) {
+      throw new TableException("SQL/Table supports only ascending time ordering")
+    }
+      
+     
+    val (offset,fetch) = (calc.offset,calc.fetch)
+    
+    //enable to extend for other types of aggregates that will not be implemented in a window
+    timeType match {
+        case _: ProcTimeType =>
+            (offset,fetch) match {
+              case (o:Any,f:Any)  => null             // offset and fetch needs retraction
+              case (_,f:Any) => null                  // offset needs retraction
+              case (o:Any,_) => null                  // fetch needs retraction
+              case _ => createSortProcTime(inputDS)   //sort can be done with/without retraction
```

Comment:
you mean specific for the order on processTime. Indeed I will check this case. Anyway for the first commit i want to target the case when there is actually some sorting - just to have the implementation also for the processFunction as a model